dartagnan/src/main/java/com/dat3m/dartagnan/program/Program.java

```diff
@@ -247,28 +254,35 @@ public BoolExpr encodeFinalRegisterValues(Context ctx){
 
             for(int i = 0; i <  events.size(); i++){
                 Event w1 = events.get(i);
-                BoolExpr lastModReg = w1.exec();
+                BooleanFormula lastModReg = w1.exec();
                 for(int j = 0; j < i; j++){
                     Event w2 = events.get(j);
                     if (!eq.areMutuallyExclusive(w1, w2)) {
-                        lastModReg = ctx.mkAnd(lastModReg, ctx.mkNot(w2.exec()));
+                        lastModReg = bmgr.and(lastModReg, bmgr.not(w2.exec()));
                     }
                 }
-                enc = ctx.mkAnd(enc, ctx.mkImplies(lastModReg,
-                        ctx.mkEq(reg.getLastValueExpr(ctx), ((RegWriter)w1).getResultRegisterExpr())));
+                BooleanFormula same = reg.getLastValueExpr(ctx) instanceof BitvectorFormula ?
+                		fmgr.getBitvectorFormulaManager().equal(
+                				(BitvectorFormula)reg.getLastValueExpr(ctx), 
+                				(BitvectorFormula)((RegWriter)w1).getResultRegisterExpr()) :
+                		fmgr.getIntegerFormulaManager().equal(
+                				(IntegerFormula)reg.getLastValueExpr(ctx), 
+                				(IntegerFormula)((RegWriter)w1).getResultRegisterExpr());
```

Comment:
It might be clever to add a Helper class that does support a generic equals method (that internally checks the types of the formulas and calls the respective managers). It seems this sort of code appears quite often.